opte/src/engine/packet.rs

```diff
@@ -1258,10 +1305,89 @@ impl Packet<Parsed> {
         self.state.body.pkt_offset
     }
 
+    /// Run the [`BodyTransform`] against this packet.
+    pub fn body_transform(
+        &mut self,
+        dir: Direction,
+        xform: &Box<dyn BodyTransform>,
+    ) -> Result<(), BodyTransformError> {
+        // We set the flag now with the assumption that the transform
+        // couild fail after modifying part of the body. In the future
+        // we could have something more sophisticated that only sets
+        // the flag if at least one byte was modified, but for now
+        // this does the job as nothing that needs top performance
+        // should make use of body transformations.
+        self.state.body_modified = true;
+
+        match self.body_segs_mut() {
+            Some(mut body_segs) => xform.run(dir, &mut body_segs),
+
+            None => {
+                self.state.body_modified = false;
+                Err(BodyTransformError::NoPayload)
+            }
+        }
+    }
+
     pub fn body_seg(&self) -> usize {
         self.state.body.seg_index
     }
 
+    /// Return a list of the body segments, or `None` if there is no
+    /// body.
+    pub fn body_segs(&self) -> Option<Vec<&[u8]>> {
+        if self.state.body.len == 0 {
+            return None;
+        }
+
+        let mut body_segs = vec![];
+        let body_seg = self.state.body.seg_index;
+
+        for (i, seg) in self.segs[body_seg..].iter().enumerate() {
+            if i == 0 {
+                body_segs.push(
+                    // Safety: The offset is based on the parse data.
+                    unsafe {
+                        seg.slice_unchecked(self.state.body.seg_offset, None)
+                    },
+                );
+            } else {
+                body_segs.push(seg.slice());
+            }
+        }
+
+        Some(body_segs)
+    }
+
+    /// Return a list of mutable body segments, or `None` if there is
+    /// no body.
+    pub fn body_segs_mut<'a>(&'a mut self) -> Option<Vec<&'a mut [u8]>> {
```

Comment:
Nit: I _think_ the lifetime can be elided here, since there's only one, but I could be wrong.